// selenium-appium/dist/index.js
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.windowsAppDriverCapabilities = exports.PageObject = exports.WebDriver2 = exports.driver = exports.By2 = void 0;
var by2_1 = require("./by2");
Object.defineProperty(exports, "By2", { enumerable: true, get: function () { return by2_1.By2; } });
var driver_1 = require("./driver");
Object.defineProperty(exports, "driver", { enumerable: true, get: function () { return driver_1.driver; } });
Object.defineProperty(exports, "WebDriver2", { enumerable: true, get: function () { return driver_1.WebDriver2; } });
var pageobject_1 = require("./pageobject");
Object.defineProperty(exports, "PageObject", { enumerable: true, get: function () { return pageobject_1.PageObject; } });
var winappdriver_1 = require("./winappdriver");
Object.defineProperty(exports, "windowsAppDriverCapabilities", { enumerable: true, get: function () { return winappdriver_1.windowsAppDriverCapabilities; } });
var config_1 = require("./config");
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return config_1.Config; } });
